### DISCUSSION TITLE
Document labels

### DISCUSSION THREAD

**Discussion by valfur03** [avitorio/outstatic#22] [Ideas]
The document labels could have multiple use cases, like adding tags to blog posts.

I think it should use a key-value pair format, so the labels can contain other kind of metadata. For example, `primary-color => red` to customize the front-end. This is a trivial example, but this is the idea.

**Comment by avitorio** [comment]
Hey @valfur03,

Like I've mentioned. This is something I definitely want to add as a feature.

My idea is to have something simpler, where the labels are just strings. But once they are added to a Document they are available to all other Documents in the same Collection. 

That's why they wouldn't be just part of the Document, but of the entire Collection.

If it's just a label we wouldn't need a key-value pair...

Thoughts?

**Reply by valfur03** [reply]
I was thinking of labels as they are implemented by Docker for example: a piece of metadata to filter things. This way, the user would be able to completely customize their use of Outstatic.

A single-value label would limit the use to categorizing the documents, according to me. But I think this would be the most used case.

Maybe the solution is to implement "labels" as single-value strings, and later on implement the feature to add "metadata", as I think of the labels right now. This would be a metadata field in the YAML front-matter, which contains a whole other YAML hierarchy. This would be a per-document configuration.

What do you think about that ?

---

Also, wouldn't it be too hard to "connect" the labels together as there is no relational database ? The only solution I see is to parse every markdown files from the collection's directory to list their labels (I don't know the project well yet, I might miss something...)

**Comment by avitorio** [comment]
Hey @valfur03 ,

Thanks for the suggestions. I believe the way to go for this is to create something that allows users to create their own fields. 
This is where I see this going.

I'll start working on a proof of concept using something like `superjson` or do it in a similar way `strapi` does it. 
You are right, there are no relational databases, but we can leverage json. 

I think one of the challenges is making sure we keep everything type safe and passing through the proper validations. Once I have something at least minimal working I'll add a link to the branch here.
